File: apis/BC/buildcraft/core/fluids/RestrictedTank.java
/*
 * Copyright (c) SpaceToad, 2011-2012
 * http://www.mod-buildcraft.com
 * 
 * BuildCraft is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package buildcraft.core.fluids;

import net.minecraftforge.fluids.Fluid;
import net.minecraftforge.fluids.FluidStack;

/**
 *
 * @author CovertJaguar <http://www.railcraft.info/>
 */
public class RestrictedTank extends Tank {

	private final Fluid[] acceptedFluids;

	public RestrictedTank(String name, int capacity, Fluid... acceptedFluids) {
		super(name, capacity);
		this.acceptedFluids = acceptedFluids;
	}

	@Override
	public int fill(FluidStack resource, boolean doFill) {
		if (!acceptsFluid(resource.getFluid()))
			return 0;
		return super.fill(resource, doFill);
	}

	public boolean acceptsFluid(Fluid fluid) {
		for (Fluid accepted : acceptedFluids) {
			if (accepted.equals(fluid))
				return true;
		}
		return false;
	}
}

File: src/main/scala/alternativemods/alteng/fluid/tank/RestrictedTank.scala
package alternativemods.alteng.fluid.tank

import net.minecraft.tileentity.TileEntity
import net.minecraftforge.fluids.{Fluid, FluidStack}
import alternativemods.alteng.util.ToolTipLine
import java.util.Locale

/**
 * No description given
 *
 * @author jk-5
 */
class RestrictedTank(val acceptedFluid: Fluid, _name: String, _capacity: Int, _owner: TileEntity) extends Tank(_name, _capacity, _owner) {
  override def fill(resource: FluidStack, doFill: Boolean): Int =
    if(!acceptsFluid(resource.getFluid)) 0 else super.fill(resource, doFill)

  def acceptsFluid(fluid: Fluid): Boolean = fluid == this.acceptedFluid

  override def refreshTooltip(){
    this.tooltip.clear()
    val fluidName = new ToolTipLine(acceptedFluid.getLocalizedName, acceptedFluid.getRarity.rarityColor)
    fluidName.spacing = 2
    tooltip.add(fluidName)
    tooltip.add(new ToolTipLine(String.format(Locale.ENGLISH, "%,d / %,d", getFluidAmount: Integer, this.getCapacity: Integer)))
  }
}
